src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -0,0 +1,271 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//-----------------------------------------------------------------------
+
+
+using System;
+using System.IO;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Globbing
+{
+    /// <summary>
+    ///     Represents a parsed MSBuild glob.
+    ///     An MSBuild glob is composed of three parts:
+    ///     - fixed directory part: "a/b/" in "a/b/**/*test*/**/*.cs"
+    ///     - wildcard directory part: "**/*test*/**/" in "a/b/**/*test*/**/*.cs"
+    ///     - file name part: "*.cs" in "a/b/**/*test*/**/*.cs"
+    /// </summary>
+    public class MsBuildGlob : IMsBuildGlob
+    {
+        //private readonly Lazy<string[]> _expandedFiles;
+        internal readonly string _fileSpec;
+        internal readonly string _globRoot;
+        private readonly string _matchFileExpression;
+        internal readonly bool _needsRecursion;
+        private readonly Lazy<Regex> _regex;
+
+        /// <summary>
+        ///     The fixed directory part.
+        /// </summary>
+        public string FixedDirectoryPart { get; }
+
+        /// <summary>
+        ///     The wildcard directory part
+        /// </summary>
+        public string WildcardDirectoryPart { get; }
+
+        /// <summary>
+        ///     The file name part
+        /// </summary>
+        public string FilenamePart { get; }
+
+        /// <summary>
+        ///     Whether the glob was parsed sucsesfully from a string.
+        ///     Illegal glob strings contain:
+        ///     - invalid path characters (other than the wildcard characters themselves)
+        ///     - "..."
+        ///     - ":"
+        ///     In addition, the wildcard directory part:
+        ///     - cannot contain ".."
+        ///     - if ** is present it must appear alone between slashes
+        /// </summary>
+        public bool IsLegal { get; }
+
+        private MsBuildGlob(
+            string globRoot,
+            string fileSpec,
+            string fixedDirectoryPart,
+            string wildcardDirectoryPart,
+            string filenamePart,
+            string matchFileExpression,
+            bool needsRecursion,
+            bool isLegalFileSpec)
+        {
+            FixedDirectoryPart = fixedDirectoryPart;
+            WildcardDirectoryPart = wildcardDirectoryPart;
+            FilenamePart = filenamePart;
+            IsLegal = isLegalFileSpec;
+
+            _globRoot = globRoot;
+            _fileSpec = fileSpec;
+            _matchFileExpression = matchFileExpression;
+            _needsRecursion = needsRecursion;
+
+            // compile the regex since it's expected to be used multiple times
+            _regex =
+                new Lazy<Regex>(
+                    () => new Regex(_matchFileExpression, FileMatcher.DefaultRegexOptions | RegexOptions.Compiled),
+                    true);
```

Comment:
Doesn't this `true` just opt into the default no-thread-safety-mode-specified case?